mailet/icalendar/src/main/java/org/apache/james/transport/mailets/ICalendarParser.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.mailet.base.GenericMailet;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import net.fortuna.ical4j.data.CalendarBuilder;
+import net.fortuna.ical4j.data.ParserException;
+import net.fortuna.ical4j.model.Calendar;
+
+/**
+ * <p>
+ * This mailet can be combined with the Strip attachment mailet.
+ * </p>
+ * <p>
+ * The ICS body part byte array arranged as map then this mailet should look for ICS and parse it with Ical4J then store it as a mail attribute
+ * </p>
+ * <p>
+ * Configuration: The mailet contains 2 mandatory attributes
+ * </p>
+ * <p>
+ *
+ * <pre>
+ *   &lt;mailet match=&quot;All&quot; class=&quot;ICalendarParser&quot; &gt;
+ *     &lt;sourceAttribute&gt;source.attribute.name&lt;/sourceAttribute&gt;  &lt;!-- The attribute which contains output value of StripAttachment mailet -- &gt;
+ *     &lt;destAttribute&gt;dest.attribute.name&lt;/destAttribute&gt;  &lt;!-- The attribute store the map of Calendar -- &gt;
+ *   &lt;/mailet &gt;
+ *
+ * </pre>
+ *
+ * </p>
+ */
+public class ICalendarParser extends GenericMailet {
+    private static final String SOURCE_ATTRIBUTE_PARAMETER_NAME = "sourceAttribute";
+    private static final String DESTINATION_ATTRIBUTE_PARAMETER_NAME = "destinationAttribute";
+
+    private static final String SOURCE_ATTRIBUTE_PARAMETER_DEFAULT_VALUE = "icsAttachments";
+    private static final String DESTINATION_ATTRIBUTE_PARAMETER_DEFAULT_VALUE = "calendars";
+
+    private String sourceAttributeName;
+    private String destinationAttributeName;
+
+    @Override
+    public void init() throws MessagingException {
+        sourceAttributeName = getInitParameter(SOURCE_ATTRIBUTE_PARAMETER_NAME, SOURCE_ATTRIBUTE_PARAMETER_DEFAULT_VALUE);
+        if (Strings.isNullOrEmpty(sourceAttributeName)) {
+            throw new MessagingException("source attribute cannot be empty");
+        }
+        destinationAttributeName = getInitParameter(DESTINATION_ATTRIBUTE_PARAMETER_NAME, DESTINATION_ATTRIBUTE_PARAMETER_DEFAULT_VALUE);
+        if (Strings.isNullOrEmpty(destinationAttributeName)) {
+            throw new MessagingException("destination attribute cannot be empty");
+        }
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Object icsAttachmentsObj = mail.getAttribute(sourceAttributeName);
+        if (icsAttachmentsObj == null || !(icsAttachmentsObj instanceof Map)) {
+            return;
+        }
+
+        Map<String, byte[]> icsAttachments = (Map<String, byte[]>) icsAttachmentsObj;
+        Map<String, Calendar> calendars = icsAttachments.entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(e -> e.getKey(), e -> createCalendar(e.getValue())))
+            .entrySet()
+            .stream()
+            .filter(optCal -> optCal.getValue().isPresent())
+            .collect(
+                Collectors.toMap(p -> p.getKey(), p -> p.getValue().get())
```

Comment:
You still collect two times, this is not efficient and it is also hard to read !

Do we need a Map as the output ? Or couldn't you use a pair ?

Eg : 

```
ics.entrySet()
    .stream()
    .map(entry -> Pair.of(entry.getKey(), createCalendar(entry.getValue())))
    .filter(pair -> pair.getValue().isPresent())
    .map(pair -> pair.of(entry.getKey(), entry.getValue().get())
    .collect(Guavate.toImmutableMap(pair -> pair.getKey(), pair -> pair.getValue());
```

If you chain collects, you have no value added by using streams....